File: src/main/java/ludwig/ide/EditorUtils.java
package ludwig.ide;

import javafx.scene.control.TextArea;

public class EditorUtils {
    public static int tokenIndex(String text, int pos) {
        if (pos >= text.length()) {
            return -1;
        }
        int index = -1;
        boolean isToken = false;
        for (int i = 0; i < Math.min(pos + 1, text.length()); i++) {
            boolean t;
            switch (text.charAt(i)) {
                case '\n':
                case '\t':
                case '\r':
                case ' ':
                case ':':
                    t = false;
                    break;
                default:
                    t = true;
                    break;
            }
            if (t && t != isToken) {
                index++;
            }
            isToken = t;
        }
        return index;
    }

    public static int tokenIndex(TextArea textArea) {
        return tokenIndex(textArea.getText(), textArea.getSelection().getStart());
    }
}

File: src/main/java/ludwig/ide/EditorUtils.kt
package ludwig.ide

import javafx.scene.control.TextArea

object EditorUtils {
    fun tokenIndex(text: String, pos: Int): Int {
        if (pos >= text.length) {
            return -1
        }
        var index = -1
        var isToken = false
        for (i in 0 until Math.min(pos + 1, text.length)) {
            val t: Boolean
            when (text[i]) {
                '\n', '\t', '\r', ' ', ':' -> t = false
                else -> t = true
            }
            if (t && t != isToken) {
                index++
            }
            isToken = t
        }
        return index
    }

    fun tokenIndex(textArea: TextArea): Int {
        return tokenIndex(textArea.text, textArea.selection.start)
    }
}
